### PR TITLE
[core] sort symbols using symbol-sort-key before placement

### LINKED ISSUE #15964: symbolSortKey does not equally sort icon and text
When using MGLSymbolStyleLayer's `symbolSortKey` expression with `textAllowsOverlap` as false, text collisions cause the lower-sort features to take priority. Features with a lower sort will draw their text and cause features of a higher sort to hide theirs.

### Steps to reproduce

 1. Draw two features on the map that overlap.
 2. Ensure they both have valid `icon` and `text` properties.
 3. Set `textOptional` to true, `iconAllowsOverlap` to true, and `textAllowsOverlap` to false.

### Expected behavior

When the features collide:
- the text and icon for the higher-sort feature should appear above the lower-sort feature's icon.
- the lower-sort feature's text should be hidden since it collides with the high-sort feature.

### Actual behavior

The icons are at the correct z-level, but the lower-sort feature's text has caused the higher-sort feature's text to hide.

The attached zip contains a sample Xcode project that demonstrates this issue. You'll just need to run a pod install and supply an API token in the app delegate. The output of the project is captured in the screenshot below. The red feature has a higher sort value, but its text is hidden.

[TextLayeringProject.zip](https://github.com/mapbox/mapbox-gl-native/files/3881697/TextLayeringProject.zip)

Screen Shot 2019-11-22 at 5 25 32 PM

If I reverse the `sort` values by multiplying them all by -1, then the red icon appears with its text, but both the text and the icon are drawn below the now higher blue icon:

Screen Shot 2019-11-22 at 5 28 10 PM

### Configuration

**Mapbox SDK versions: 5.4.0** 
**iOS/macOS versions: 13.2.2** 
**Device/simulator models: 11**
**Xcode version: 11.2.1**

### PR DESCRIPTION
**Pull request by ansis**
fix #15964
partially port https://github.com/mapbox/mapbox-gl-js/pull/9054

Currently symbol-sort-key works:
- across tile boundaries with -allow-overlap: true
- within tiles for collision detection

This pr makes it work for collisions across tile boundaries as well.

This splits the bucket into ranges of symbolInstances with the same sort key. All the ranges from all the tiles in a layer are inserted into a sorted list. Placement is done in the order of the list.

@pozdnyakov feel free to assign reviewing to someone else if you want

### CODE REVIEW

File: src/mbgl/text/placement.hpp [hunk 1]
```diff
@@ -87,11 +87,18 @@ class CollisionGroups {
 
 class BucketPlacementParameters {
 public:
+    friend bool operator<(const BucketPlacementParameters& lhs, const BucketPlacementParameters& rhs) {
+        return lhs.sortKey < rhs.sortKey;
+    }
+    SymbolBucket& bucket;
```
Comment (pozdnyakov) [inline src/mbgl/text/placement.hpp]: `BucketPlacementParameters` shall not know about `SymbolBucket`

File: src/mbgl/text/placement.hpp [hunk 2]
```diff
@@ -87,11 +87,18 @@ class CollisionGroups {
...
     const RenderTile& tile;
     const mat4& projMatrix;
     std::string sourceId;
     std::shared_ptr<FeatureIndex> featureIndex;
     bool showCollisionBoxes;
+    float sortKey;
+    size_t symbolInstanceStart;
+    size_t symbolInstanceEnd;
```
Comment (pozdnyakov) [inline src/mbgl/text/placement.hpp]: could the newly added fields be `SymbolBucket` properties?

File: src/mbgl/text/placement.hpp (hunk 1 above)
Comment (ansis) [reply src/mbgl/text/placement.hpp]: Would it be ok for `BucketPlacementParameters` to have a reference to `Bucket`?

Comment (pozdnyakov) [reply src/mbgl/text/placement.hpp]: Could we instead just pre-sort `LayerPlacementData` in [RenderSymbolLayer::prepare()](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/renderer/layers/render_symbol_layer.cpp#L571) accordingly to the the symbol bucket sort key?

File: src/mbgl/renderer/render_layer.hpp [hunk 3]
```diff
@@ -29,7 +30,7 @@ class LayerRenderData {
 
 class LayerPlacementData {
 public:
-    std::reference_wrapper<Bucket> bucket;
+    std::reference_wrapper<SymbolBucket> bucket;
```
[the remark this answers is not in the document]
Comment (ansis) [reply src/mbgl/renderer/render_layer.hpp]: Right, thanks! I switched back to `Bucket`

File: src/mbgl/text/placement.hpp (hunk 1 above)
Comment (ansis) [reply src/mbgl/text/placement.hpp]: Yep, changed to your suggestion. This means that there are now possibly multiple entries into LayerPlacementData per tile which we have to account for elsewhere

File: src/mbgl/text/placement.hpp (hunk 2 above)
Comment (ansis) [reply src/mbgl/text/placement.hpp]: These properties come from `SortKeyRange`. A bucket can have many `SortKeyRange`s

**Comment by ansis** [comment]
@pozdnyakov sorry about the delay getting back to this! I made changes based on your review so this is ready for another look.

The `android-api-breakage` ci run is failing. Any ideas why?

File: src/mbgl/renderer/layers/render_symbol_layer.cpp [hunk 4]
```diff
@@ -580,7 +582,32 @@ void RenderSymbolLayer::prepare(const LayerPrepareParameters& params) {
             const Tile* tile = params.source->getRenderedTile(renderTile.id);
             assert(tile);
             assert(tile->kind == Tile::Kind::Geometry);
-            placementData.push_back({*bucket, renderTile, static_cast<const GeometryTile*>(tile)->getFeatureIndex()});
+
+            bool firstInBucket = true;
+
+            if (bucket->sortKeyRanges.size() == 0) {
```
Comment (pozdnyakov) [inline src/mbgl/renderer/layers/render_symbol_layer.cpp]: nit: `bucket->sortKeyRanges.empty()`

File: src/mbgl/renderer/layers/render_symbol_layer.cpp [hunk 5]
```diff
@@ -580,7 +582,32 @@ void RenderSymbolLayer::prepare(const LayerPrepareParameters& params) {
...
+                                         static_cast<const GeometryTile*>(tile)->getFeatureIndex(),
+                                         firstInBucket,
+                                         0.0f,
+                                         0,
+                                         bucket->symbolInstances.size()});
+            } else {
+                for (const SortKeyRange& sortKeyRange : bucket->sortKeyRanges) {
+                    LayerPlacementData layerData{*bucket,
+                                                 renderTile,
+                                                 static_cast<const GeometryTile*>(tile)->getFeatureIndex(),
```
Comment (pozdnyakov) [inline src/mbgl/renderer/layers/render_symbol_layer.cpp]: nit: might make sense to cache `static_cast<const GeometryTile*>(tile)->getFeatureIndex()` in a local variable

File: src/mbgl/text/placement.cpp [hunk 6]
```diff
@@ -656,18 +661,17 @@ void Placement::placeBucket(const SymbolBucket& bucket,
         }
 
     } else {
-        for (const SymbolInstance& symbol : bucket.symbolInstances) {
-            placeSymbol(symbol);
+        auto endIt = bucket.symbolInstances.begin() + params.symbolInstanceEnd;
```
Comment (pozdnyakov) [inline src/mbgl/text/placement.cpp]: do we check that `endIt <= bucket.symbolInstances.end()`?

File: src/mbgl/layout/symbol_layout.cpp [hunk 7]
```diff
@@ -572,23 +572,47 @@ void SymbolLayout::addFeature(const std::size_t layoutFeatureIndex,
...
+    auto addSymbolInstance = [&](Anchor& anchor, std::shared_ptr<SymbolInstanceSharedData> sharedData) {
         assert(sharedData);
-        const bool anchorInsideTile = anchor.point.x >= 0 && anchor.point.x < util::EXTENT && anchor.point.y >= 0 && anchor.point.y < util::EXTENT;
+        const bool anchorInsideTile = anchor.point.x >= 0 && anchor.point.x < util::EXTENT && anchor.point.y >= 0 &&
+                                      anchor.point.y < util::EXTENT;
 
         if (mode == MapMode::Tile || anchorInsideTile) {
+            if (sortFeaturesByKey) {
+                if (sortKeyRanges.size() && sortKeyRanges.back().sortKey == feature.sortKey) {
+                    sortKeyRanges.back().symbolInstanceEnd = symbolInstances.size() + 1;
```
Comment (pozdnyakov) [inline src/mbgl/layout/symbol_layout.cpp]: `sortKeyRanges.back().symbolInstanceEnd = symbolInstances.size();` shall be enough, no? `begin()` + size = `end()` :thinking:

File: src/mbgl/text/placement.cpp [hunk 8]
```diff
@@ -656,18 +661,17 @@ void Placement::placeBucket(const SymbolBucket& bucket,
...
+        for (auto it = bucket.symbolInstances.begin() + params.symbolInstanceStart; it != endIt; it++) {
```
Comment (pozdnyakov) [inline src/mbgl/text/placement.cpp]: styling nit: better have `auto beginIt = bucket.symbolInstances.begin() + params.symbolInstanceStar`

Comment (pozdnyakov) [reply src/mbgl/text/placement.cpp]: nit: ++it

File: src/mbgl/layout/symbol_layout.cpp (hunk 7 above)
Comment (ansis) [reply src/mbgl/layout/symbol_layout.cpp]: It looks like I'm setting it before adding the new symbol instance. I guess moving it after and removing the `+ 1` would be cleaner

File: src/mbgl/text/placement.cpp (hunk 6 above)
Comment (ansis) [reply src/mbgl/text/placement.cpp]: Nope, would an assert be what you're looking for?

Comment (pozdnyakov) [reply src/mbgl/text/placement.cpp]: assert is fine (as long as there are no possible code paths that could violate it :smile: )

**Pull request merged**